src/common/tim.rs

```diff
@@ -0,0 +1,71 @@
+use std::{
+    sync::Arc,
+    time::{Duration, Instant}
+};
+
+use crate::rt::{Interval, Sleep, Timer};
+
+/// A user-provided timer to time background tasks.
+pub(crate) type Tim = Option<Arc<dyn Timer + Send + Sync>>;
+
+/*
+pub(crate) fn timeout<F>(tim: Tim, future: F, duration: Duration) -> HyperTimeout<F> {
+    HyperTimeout { sleep: tim.sleep(duration), future: future }
+}
+
+pin_project_lite::pin_project! {
+    pub(crate) struct HyperTimeout<F> {
+        sleep: Box<dyn Sleep>,
+        #[pin]
+        future: F
+    }
+}
+
+pub(crate) struct Timeout;
+
+impl<F> Future for HyperTimeout<F> where F: Future {
+
+    type Output = Result<F::Output, Timeout>;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output>{
+        let mut this = self.project();
+        if let Poll::Ready(v) = this.future.poll(ctx) {
+            return Poll::Ready(Ok(v));
+        }
+
+        if let Poll::Ready(_) = Pin::new(&mut this.sleep).poll(ctx) {
+            return Poll::Ready(Err(Timeout));
+        }
+
+        return Poll::Pending;
+    }
+}
+*/
+
+impl Timer for Tim {
+    fn sleep(&self, duration: Duration) -> Box<dyn Sleep + Unpin> {
+        match *self {
+            None => {
+                panic!("You must supply a timer.")
```

Comment:
On the second point, my original reasoning was that if you've tried set up something like header reading timeouts, but we can't do it because you didn't give us a timer, we should fail loudly (panic). The failure mode otherwise--not doing the timeouts--seems subtle and likely only to be discovered when it causes a problem for the user in the future.

On the first point, I'm not sure I understand your suggestion fully. You're saying that instead of implementing `Timer` on `Time`, we could add these as inherent methods of `Timer`? Would you mind elaborating a bit more?